x/pubkey/keeper/endblock.go

```diff
@@ -0,0 +1,76 @@
+package keeper
+
+import (
+	"errors"
+
+	"cosmossdk.io/collections"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/sedaprotocol/seda-chain/app/utils"
+)
+
+func (k Keeper) EndBlock(ctx sdk.Context) (err error) {
+	// Use defer to prevent returning an error, which would cause
+	// the chain to halt.
+	defer func() {
+		// Handle a panic.
+		if r := recover(); r != nil {
+			k.Logger(ctx).Error("recovered from panic in pubkey end block", "err", r)
+		}
+		// Handle an error.
+		if err != nil {
+			k.Logger(ctx).Error("error in pubkey end block", "err", err)
+		}
+		err = nil
+	}()
+
+	// If secp256k1 proving scheme is already enabled, do nothing.
+	isEnabled, err := k.IsProvingSchemeEnabled(ctx, utils.SEDAKeyIndexSecp256k1)
+	if err != nil {
+		return err
+	}
+	if isEnabled {
+		return
+	}
+
+	// If the sum of the voting power has reached 80%, enable secp256k1
+	// proving scheme.
+	totalPower, err := k.stakingKeeper.GetLastTotalPower(ctx)
```

Comment:
We might want to add a delay here so validators have a chance not to get jailed. So once we hit 80% we wait X blocks before enabling it.

Reply:
Added activation height to the proving scheme type (added explanation in the prompt)